packages/gatsby/src/schema/node-model.js

```
@@ -307,35 +320,75 @@ class LocalNodeModel {
-      result.forEach(node => this.trackInlineObjectsInRootNode(node))
-    }
+    entries.forEach(node => this.trackInlineObjectsInRootNode(node))
 
     if (trackInlineObjectsActivity) {
       trackInlineObjectsActivity.end()
     }
+    return { gqlType, entries, totalCount }
+  }
+
+  /**
+   * Get nodes of a type matching the specified query.
+   *
+   * Note: this method returns a slice of result when `skip` and `limit` are set.
+   *
+   * @param {Object} args
+   * @param {Object} args.query Query arguments (`filter`, `sort`, `skip`, `limit`)
+   * @param {(string|GraphQLOutputType)} args.type Type
+   * @param {PageDependencies} [pageDependencies]
+   * @returns {Promise<IQueryResult>}
+   */
+  async findAll(args, pageDependencies = {}) {
+    const { gqlType, ...result } = await this._query(args, pageDependencies)
 
     // Tracking connections by default:
-    if (!firstOnly && typeof pageDependencies.connectionType === `undefined`) {
+    if (typeof pageDependencies.connectionType === `undefined`) {
       pageDependencies.connectionType = gqlType.name
     }
-    return this.trackPageDependencies(result, pageDependencies)
+    this.trackPageDependencies(result.entries, pageDependencies)
+    return result
+  }
+
+  /**
+   * Get the first node of a type matching the specified query.
+   *
+   * @param {Object} args
+   * @param {Object} args.query Query arguments (supports: `filter`)
+   * @param {(string|GraphQLOutputType)} args.type Type
+   * @param {PageDependencies} [pageDependencies]
+   * @returns {Promise<Node | null>}
+   */
+  async findOne(args, pageDependencies = {}) {
+    const { query } = args
+    if (query.sort?.fields?.length > 0) {
+      // If we support sorting and return the first node based on sorting
+      // we'll have to always track connection not an individual node
+      reporter.warn(
+        `nodeModel.findOne() does not support sorting. Use nodeModel.findAll({ query: { limit: 1 } }) instead`
+      )
```

A subtle nuance. Right now you can do this in your custom resolver:

```js
const firstNode = await nodeModel.runQuery({
  query: { sort: { fields: [`foo`] } }
  firstOnly: true
})
```

But it will silently ignore sorting under the hood. So you won't get "the first node sorted by `foo`". After this change it will fail with this error. Not sure if we should consider it a breaking change or not - I can remove it for now and keep silently ignoring sort.

---

why can't we run findAll for our users ourselves? As a developer I would find it weird to do a findAll(limit: 1) instead of using findOne

---

`findOne` registers data dependency for a single node. `findAll` registers data dependency for the whole connection. So in the first case query will be invalidated only when this specific node changes but in the second case - when _any_ node of this type changes.

We want to have as many individual node dependencies but not connection dependencies. So ideally if you don't care about sorting - `findOne` is the best choice for warm / inc.builds.

---

Ah now you're comment makes sense to me.